actor-macros/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Attribute, DeriveInput, GenericParam, Lit};
+
+#[proc_macro_derive(Actor, attributes(actor))]
+pub fn actor(item: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(item as DeriveInput);
+    // Extract struct name
+    let struct_name = item.ident;
+
+    let Config {
+        trait_name,
+        envelope_name,
+        handle_name,
+    } = Config::from_attributes(&struct_name, &item.attrs);
+
+    let generics = item.generics;
+    let generic_types: Vec<_> = generics
+        .params
+        .iter()
+        .filter_map(|param| match param {
+            GenericParam::Type(type_param) => Some(&type_param.ident),
+            _ => None,
+        })
+        .collect();
+    let phantom_fields = generic_types.iter().map(|ty| {
+        let name = syn::Ident::new(&format!("__{}", ty).to_lowercase(), ty.span());
+        quote! {
+            #name: std::marker::PhantomData<#ty>
+        }
+    });
+    let phantom_values: Vec<_> = generic_types
+        .iter()
+        .map(|ty| {
+            let name = syn::Ident::new(&format!("__{}", ty).to_lowercase(), ty.span());
+            quote! {
+                #name: Default::default()
+            }
+        })
+        .collect();
+
+    // Generate the implementation
+    let expanded = quote! {
+        impl #generics ceramic_actor::Actor for #struct_name < #(#generic_types,)*> {
+            type Envelope = #envelope_name;
+        }
+        impl #generics #trait_name for #struct_name < #(#generic_types,)*> { }
+
+        impl #generics #struct_name < #(#generic_types,)*> {
+            /// Start the actor returning a handle that can be easily cloned and shared.
+            /// The actor stops once all handles are dropped.
+            pub fn spawn(size: usize, actor: impl #trait_name + ::std::marker::Send + 'static, shutdown: impl ::std::future::Future<Output=()> + ::std::marker::Send + 'static) -> (#handle_name < #(#generic_types,)*>, tokio::task::JoinHandle<()>) {
```

Comment:
do we care to return the `JoinHandle` here? It seems like we'll pretty much just drop it and rely on the actor Handle to control the task. Not really a big deal, maybe it's more important in the future usage.

Reply:
The examples do not use it, but the pipeline code does as well as test code.